File: src/main/java/me/zeroeightsix/kami/module/modules/combat/Aura.java
package me.zeroeightsix.kami.module.modules.combat;

import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.module.modules.misc.AutoTool;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.EntityUtil;
import me.zeroeightsix.kami.util.Friends;
import me.zeroeightsix.kami.util.LagCompensator;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Items;
import net.minecraft.item.ItemAxe;
import net.minecraft.item.ItemSword;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.Vec3d;

/**
 * Created by 086 on 12/12/2017.
 * Updated by hub on 31 October 2019
 * Updated by S-B99 on 10/04/20
 * Updated by bot-debug on 10/04/20
 */
@Module.Info(name = "Aura", category = Module.Category.COMBAT, description = "Hits entities around you")
public class Aura extends Module {
    private Setting<WaitMode> delayMode = register(Settings.e("Mode", WaitMode.DELAY));
    private Setting<Boolean> autoSpamDelay = register(Settings.booleanBuilder("Auto Spam Delay").withValue(true).withVisibility(v -> delayMode.getValue().equals(WaitMode.SPAM)).build());
    private Setting<Double> waitTick = register(Settings.doubleBuilder("Spam Delay").withMinimum(0.1).withValue(2.0).withMaximum(20.0).withVisibility(v -> !autoSpamDelay.getValue() && delayMode.getValue().equals(WaitMode.SPAM)).build());
    private Setting<Boolean> eat = register(Settings.b("While Eating", true));
    private Setting<Boolean> multi = register(Settings.b("Multi", true));
    private Setting<Boolean> attackPlayers = register(Settings.b("Players", true));
    private Setting<Boolean> attackMobs = register(Settings.b("Mobs", false));
    private Setting<Boolean> attackAnimals = register(Settings.b("Animals", false));
    private Setting<Double> hitRange = register(Settings.d("Hit Range", 5.5d));
    private Setting<Boolean> ignoreWalls = register(Settings.b("Ignore Walls", true));
    private Setting<HitMode> prefer = register(Settings.e("Prefer", HitMode.SWORD));
    private Setting<Boolean> autoTool = register(Settings.b("Auto Weapon", true));
    private Setting<Boolean> sync = register(Settings.b("TPS Sync", false));
    private Setting<Boolean> infoMsg = register(Settings.booleanBuilder("Info Message").withValue(true).withVisibility(v -> false));

    private int waitCounter;

    public enum HitMode { SWORD, AXE, NONE }
    private enum WaitMode { DELAY, SPAM }

    @Override
    public void onEnable() {
        if (mc.player == null) return;
        if (autoSpamDelay.getValue() && infoMsg.getValue()) {
            infoMsg.setValue(false);
            Command.sendWarningMessage("[Aura] When Auto Spam Delay is turned on whatever you give Spam Delay doesn't matter");
        }
    }

    @Override
    public void onUpdate() {
        if (mc.player == null || mc.player.isDead) return;

        float autoWaitTick = 20.0f - LagCompensator.INSTANCE.getTickRate();
        final boolean canAttack = (mc.player.getCooledAttackStrength(this.sync.getValue() ? -autoWaitTick : 0.0f) >= 1);

        if (!eat.getValue()) {
            boolean shield = mc.player.getHeldItemOffhand().getItem().equals(Items.SHIELD) && mc.player.getActiveHand() == EnumHand.OFF_HAND;

            if (mc.player.isHandActive() && !shield) {
                return;
            }
        }
        if (delayMode.getValue().equals(WaitMode.DELAY)) {
            if (mc.player.getCooledAttackStrength(getLagComp()) < 1) {
                return;
            } else if (mc.player.ticksExisted % 2 != 0) {
                return;
            }
        }

        if (autoSpamDelay.getValue()) {
            if (delayMode.getValue().equals(WaitMode.SPAM) && autoWaitTick > 0) {
                if (sync.getValue()) {
                    if (waitCounter < autoWaitTick) {
                        waitCounter++;
                        return;
                    } else {
                        waitCounter = 0;
                    }
                } else {
                    if (!canAttack) return;
                }
            }
        } else {
            if (delayMode.getValue().equals(WaitMode.SPAM) && waitTick.getValue() > 0) {
                if (waitCounter < waitTick.getValue()) {
                    waitCounter++;
                    return;
                } else {
                    waitCounter = 0;
                }
            }
        }
        if ((autoTool.getValue())) {
            AutoTool.equipBestWeapon(prefer.getValue());
        }
        for (Entity target : mc.world.loadedEntityList) {
            if (!EntityUtil.isLiving(target))
                continue;
            if (target == mc.player)
                continue;
            if (mc.player.getDistance(target) > hitRange.getValue())
                continue;
            if (((EntityLivingBase) target).getHealth() <= 0)
                continue;
            if (delayMode.getValue().equals(WaitMode.DELAY) && ((EntityLivingBase) target).hurtTime != 0)
                continue;
            if (!ignoreWalls.getValue() && (!mc.player.canEntityBeSeen(target) && !canEntityFeetBeSeen(target)))
                continue; // If walls is on & you can't see the feet or head of the target, skip. 2 raytraces needed

            if (attackPlayers.getValue() && target instanceof EntityPlayer && !Friends.isFriend(target.getName())) {
                attack(target);
                if (!multi.getValue()) return;
            } else {
                if (EntityUtil.isPassive(target) ? attackAnimals.getValue() : (EntityUtil.isMobAggressive(target) && attackMobs.getValue())) {
                    attack(target);
                    if (!multi.getValue()) return;
                }
            }
        }
    }

    private void attack(Entity e) {
        mc.playerController.attackEntity(mc.player, e);
        mc.player.swingArm(EnumHand.MAIN_HAND);
    }

    private float getLagComp() {
        if (delayMode.getValue().equals(WaitMode.DELAY)) {
            return -(20 - LagCompensator.INSTANCE.getTickRate());
        }
        return 0.0F;
    }

    private boolean canEntityFeetBeSeen(Entity entityIn) {
        return mc.world.rayTraceBlocks(new Vec3d(mc.player.posX, mc.player.posY + mc.player.getEyeHeight(), mc.player.posZ), new Vec3d(entityIn.posX, entityIn.posY, entityIn.posZ), false, true, false) == null;
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/Aura.kt
package me.zeroeightsix.kami.module.modules.combat

import me.zeroeightsix.kami.manager.mangers.CombatManager
import me.zeroeightsix.kami.manager.mangers.PlayerPacketManager
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Settings
import me.zeroeightsix.kami.util.LagCompensator
import me.zeroeightsix.kami.util.combat.CombatUtils
import me.zeroeightsix.kami.util.math.RotationUtils
import me.zeroeightsix.kami.util.math.RotationUtils.faceEntityClosest
import me.zeroeightsix.kami.util.math.Vec2f
import net.minecraft.entity.Entity
import net.minecraft.util.EnumHand

@Module.Info(
        name = "Aura",
        category = Module.Category.COMBAT,
        description = "Hits entities around you",
        modulePriority = 50
)
object Aura : Module() {
    private val delayMode = register(Settings.e<WaitMode>("Mode", WaitMode.DELAY))
    private val multi = register(Settings.b("Multi", false))
    private val lockView = register(Settings.booleanBuilder("LockView").withValue(false).withVisibility { !multi.value })
    private val spoofRotation = register(Settings.booleanBuilder("SpoofRotation").withValue(true).withVisibility { !multi.value && !lockView.value })
    private val waitTick = register(Settings.floatBuilder("SpamDelay").withMinimum(0.1f).withValue(2.0f).withMaximum(40.0f).withVisibility { delayMode.value == WaitMode.SPAM })
    val range = register(Settings.floatBuilder("Range").withValue(5f).withRange(0f, 10f))
    private val tpsSync = register(Settings.b("TPSSync", false))
    private val autoTool = register(Settings.b("AutoWeapon", true))
    private val prefer = register(Settings.enumBuilder(CombatUtils.PreferWeapon::class.java).withName("Prefer").withValue(CombatUtils.PreferWeapon.SWORD).withVisibility { autoTool.value })
    private val disableOnDeath = register(Settings.b("DisableOnDeath", false))

    private var inactiveTicks = 0
    private var tickCount = 0

    private enum class WaitMode {
        DELAY, SPAM
    }

    override fun isActive(): Boolean {
        return inactiveTicks <= 20 && isEnabled
    }

    override fun onUpdate() {
        inactiveTicks++
        if (mc.player.isDead) {
            if (mc.player.isDead && disableOnDeath.value) disable()
            return
        }
        if (!CombatManager.isOnTopPriority(this) || CombatSetting.pause) return
        val targetList = CombatManager.targetList
        val target = CombatManager.target

        if (multi.value && targetList.isNotEmpty()) {
            inactiveTicks = 0
            if (canAttack()) {
                for (entity in targetList) {
                    if (mc.player.getDistance(entity) > range.value) continue
                    attack(entity)
                }
            }
        } else if (target != null && mc.player.getDistance(target) < range.value) {
            inactiveTicks = 0
            if (lockView.value) {
                faceEntityClosest(target)
            } else if (spoofRotation.value) {
                val rotation = Vec2f(RotationUtils.getRotationToEntityClosest(target))
                PlayerPacketManager.addPacket(this, PlayerPacketManager.PlayerPacket(rotating = true, rotation = rotation))
            }
            if (canAttack()) attack(target)
        }
    }

    private fun canAttack(): Boolean {
        val adjustTicks = if (!tpsSync.value) 0f else LagCompensator.adjustTicks
        return if (delayMode.value == WaitMode.DELAY) {
            (mc.player.getCooledAttackStrength(adjustTicks) >= 1f)
        } else {
            if (tickCount < waitTick.value) {
                tickCount++
                false
            } else {
                tickCount = 0
                true
            }
        }
    }

    private fun attack(e: Entity) {
        if (autoTool.value) CombatUtils.equipBestWeapon(prefer.value as CombatUtils.PreferWeapon)
        mc.playerController.attackEntity(mc.player, e)
        mc.player.swingArm(EnumHand.MAIN_HAND)
    }
}